File: app/src/main/java/com/fjun/hassalarm/BanActivity.java
package com.fjun.hassalarm;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.fjun.hassalarm.databinding.ActivityBanlistBinding;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

import static com.fjun.hassalarm.Constants.KEY_IGNORED_PACKAGES;
import static com.fjun.hassalarm.Constants.PREFS_NAME;

public class BanActivity extends AppCompatActivity {

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        com.fjun.hassalarm.databinding.ActivityBanlistBinding mBinding = ActivityBanlistBinding.inflate(getLayoutInflater());
        setContentView(mBinding.root);

        final AlarmManager alarmManager = getSystemService(AlarmManager.class);
        final AlarmManager.AlarmClockInfo alarmClockInfo = alarmManager.getNextAlarmClock();

        final String missingPackageName = getString(R.string.ban_no_alarm_set);
        final PendingIntent pendingIntent = alarmClockInfo != null ? alarmClockInfo.getShowIntent() : null;
        final String pendingPackageName = pendingIntent != null ? pendingIntent.getCreatorPackage() : missingPackageName;
        mBinding.add.setEnabled(pendingIntent != null);

        mBinding.nextPackage.setText(pendingPackageName != null ? pendingPackageName : missingPackageName);

        final SharedPreferences sharedPreferences = getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);

        mBinding.list.setLayoutManager(new LinearLayoutManager(this));
        final BanAdapter adapter = new BanAdapter((packageName, adapter1) -> {
            // REMOVE
            final Set<String> set = sharedPreferences.getStringSet(KEY_IGNORED_PACKAGES, new HashSet<>());
            set.remove(packageName);
            save(sharedPreferences, set);
            setList(sharedPreferences, adapter1);
        });
        mBinding.list.setAdapter(adapter);

        mBinding.add.setOnClickListener(
                v -> {
                    // ADD
                    if (pendingIntent != null) {
                        final Set<String> set =
                                sharedPreferences.getStringSet(KEY_IGNORED_PACKAGES, new HashSet<>());
                        set.add(pendingIntent.getCreatorPackage());
                        save(sharedPreferences, set);
                        setList(sharedPreferences, adapter);
                    }
                });

        // LOAD
        setList(sharedPreferences, adapter);
    }

    private static void setList(SharedPreferences sharedPreferences, BanAdapter adapter) {
        adapter.set(new ArrayList<>(sharedPreferences.getStringSet(KEY_IGNORED_PACKAGES, new HashSet<>())));
    }

    private static void save(SharedPreferences sharedPreferences, Set<String> set) {
        sharedPreferences.edit().putStringSet(KEY_IGNORED_PACKAGES, set).apply();
    }

    public static Intent createIntent(Context context) {
        return new Intent(context, BanActivity.class);
    }
}

File: app/src/main/java/com/fjun/hassalarm/BanActivity.kt
package com.fjun.hassalarm

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.fjun.hassalarm.databinding.ActivityBanlistBinding
import java.util.regex.Pattern

class BanActivity : AppCompatActivity() {

    private lateinit var adapter: BanAdapter
    private lateinit var sharedPreferences: SharedPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding = ActivityBanlistBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setupInsets(binding.root)

        val alarmManager = getSystemService(AlarmManager::class.java)
        val alarmClockInfo = alarmManager.nextAlarmClock
        val missingPackageName = getString(R.string.ban_no_alarm_set)
        val pendingIntent = alarmClockInfo?.showIntent
        val pendingPackageName =
            if (pendingIntent != null) {
                pendingIntent.creatorPackage
            } else {
                missingPackageName
            }

        sharedPreferences = getSharedPreferences(PREFS_NAME, MODE_PRIVATE)
        adapter = BanAdapter { packageName: String ->
            // REMOVE
            val set = sharedPreferences
                .getStringSet(KEY_IGNORED_PACKAGES, HashSet<String>()) ?: setOf()
            val newSet = set.toMutableSet().apply {
                remove(packageName)
            }
            save(newSet)
            setList()
        }

        with(binding) {
            add.isEnabled = pendingIntent != null
            nextPackage.text = pendingPackageName ?: missingPackageName
            list.layoutManager = LinearLayoutManager(this@BanActivity)
            list.adapter = adapter
            add.setOnClickListener {
                addNextAlarmPackage(pendingIntent)
            }
            addManual.isEnabled = false
            manualPackageEditText.addTextChangedListener(object : TextWatcher {
                override fun afterTextChanged(s: Editable?) {
                    addManual.isEnabled = isValidPackageName(s.toString())
                }

                override fun beforeTextChanged(
                    s: CharSequence?,
                    start: Int,
                    count: Int,
                    after: Int
                ) {
                }

                override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            })
            addManual.setOnClickListener {
                val manualPackage = manualPackageEditText.text.toString()
                addPackage(manualPackage)
                manualPackageEditText.text = null
            }
        }

        setList()
    }

    private fun isValidPackageName(packageName: String) =
        Pattern.matches("^[a-zA-Z_][a-zA-Z0-9_]*(\\.[a-zA-Z_][a-zA-Z0-9_]*)+$", packageName)

    private fun addNextAlarmPackage(pendingIntent: PendingIntent?) {
        val creatorPackage = pendingIntent?.creatorPackage
        if (!creatorPackage.isNullOrEmpty()) {
            addPackage(creatorPackage)
        }
    }

    private fun addPackage(packageName: String) {
        if (packageName.isNotEmpty()) {
            val set = sharedPreferences
                .getStringSet(KEY_IGNORED_PACKAGES, HashSet<String>()) ?: setOf()
            val newSet = set.toMutableSet().apply {
                add(packageName)
            }
            save(newSet)
            setList()
        }
    }

    private fun setList() {
        adapter.set(
            (sharedPreferences.getStringSet(KEY_IGNORED_PACKAGES, HashSet<String>())
                ?: setOf()).toList()
        )
    }

    private fun save(set: Set<String>) {
        sharedPreferences.edit().putStringSet(KEY_IGNORED_PACKAGES, set)
            .apply()
    }

    companion object {
        fun createIntent(context: Context?): Intent = Intent(context, BanActivity::class.java)
    }
}